woocommerce/woocommerce-android
# [Hack Week] Refactor notifications logic to allow deleting and re-creating notification channels

Part of: #10327


### Description
This is a first part of implementing #10327, the changes in this PR are:
1. Create a component that handles the creation of notification channels.
2. Update the notification logic components to use the above class.
3. Few improvements to the logic and tests.
4. Add logic that handles deleting and re-creating notification channels.

### Testing instructions
1. Apply the patch from below.
2. Open the notification settings screen for the app from the system settings.
3. Choose the "new order" channel, and make it silent.
4. Open the app (make sure your are signed in)
5. Go back to the notification settings, and open the "new order" channel a second time.
6. Confirm that it restored the "default" importance, and it's not silent anymore.
7. Test a push notification, and confirm it uses the cha-ching sound.

```patch
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt	(revision 938d20dbb2df107ee034e1ac0176fb95deb3274c)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt	(date 1701705018076)
@@ -56,6 +56,8 @@
 import com.woocommerce.android.extensions.hide
 import com.woocommerce.android.extensions.navigateSafely
 import com.woocommerce.android.model.Notification
+import com.woocommerce.android.notifications.NotificationChannelType
+import com.woocommerce.android.notifications.NotificationChannelsHandler
 import com.woocommerce.android.support.help.HelpActivity
 import com.woocommerce.android.support.help.HelpOrigin
 import com.woocommerce.android.tools.SelectedSite
@@ -165,6 +167,7 @@
     @Inject lateinit var appWidgetUpdaters: WidgetUpdater.StatsWidgetUpdaters
     @Inject lateinit var trialStatusBarFormatterFactory: TrialStatusBarFormatterFactory
     @Inject lateinit var startUpgradeFlowFactory: StartUpgradeFlowFactory
+    @Inject lateinit var notificationChannelsHandler: NotificationChannelsHandler
 
     private val viewModel: MainActivityViewModel by viewModels()
 
@@ -354,6 +357,10 @@
 
         checkConnection()
         viewModel.showFeatureAnnouncementIfNeeded()
+
+        if (!notificationChannelsHandler.checkNotificationChannelSound(NotificationChannelType.NEW_ORDER)) {
+            notificationChannelsHandler.recreateNotificationChannel(NotificationChannelType.NEW_ORDER)
+        }
     }
 
     override fun onPause() {
```
